### PR TITLE
deployment: document use of snap DPL provider

### PR DESCRIPTION
**Pull request by kyrofa**
The previous docs discussed using `snapcraft enable-ci`. The snap DPL provider is a much cleaner way of doing this.

### CODE REVIEW

File: user/deployment/snaps.md [hunk 1]
```diff
@@ -1,38 +1,69 @@
...
-title: Ubuntu Snap Store
+title: Snap Store
 layout: en
 permalink: /user/deployment/snaps/
 ---
 
-Travis CI can automate the continuous delivery to the [Snaps](https://snapcraft.io) Store.
+Travis CI can automatically upload and release your built snap to the [snap store](https://snapcraft.io) after a successful build.
 
 Snaps are a new packaging and delivery system for Ubuntu, Debian, Fedora, openSuse and other [Linux distros](https://snapcraft.io/docs/core/install). Snaps are secure, they bundle all their dependencies and they are designed to be part of the upstream development workflow in a seamless way. This means that no distro maintainers are involved, and that the delivery to the store can be fully automated. The snaps installed in the users' machines are auto-updated, so a few hours after you push a new snap to the store, all your users will get the most recent version.
```
Comment (evandandrea) [inline user/deployment/snaps.md]: Can we rewrite this while we're here? How about:

>[Snapcraft](https://snapcraft.io) lets you distribute to all Ubuntu releases and a [growing set of Linux distributions](https://docs.snapcraft.io/core/install) with a single artefact. You publish and update at your pace while still reaching everyone; you’re not locked to the release cycle of Ubuntu or any other distribution. The updates apply automatically and roll back if anything goes wrong. They’re secure; each update is cryptographically signed and is tamper-proof once installed. The applications are locked down using the same container primitives found in Docker and LXD.

File: user/deployment/snaps.md [hunk 2]
```diff
@@ -1,38 +1,69 @@
...
 
-To automate continuous delivery of snaps to the Ubuntu store:
+All it takes is the following in your `.travis.yml`:
```
Comment (evandandrea) [inline user/deployment/snaps.md]: To upload your snap, add the following to your `.travis.yml`:

File: user/deployment/snaps.md [hunk 3]
```diff
@@ -1,38 +1,69 @@
...
-3.  Run `snapcraft` and `snapcraft push` in your `.travis.yml`:
+```yaml
+deploy:
+  provider: snap
+  snap: my.snap
+  channel: edge
+```
+{: data-file=".travis.yml"}
+
+In order for Travis CI to have permission to upload and release the snap on your behalf, you need to export a login token by installing Snapraft:
```
Comment (evandandrea) [inline user/deployment/snaps.md]: Snapcraft the platform is capitalised, snapcraft the CLI tool is not.

How about:
>In order for Travis CI to have permission to upload the snap on your behalf, you need to export a Snap Store login token. You can do this with the snapraft command-line tool, once you have \[enabled snap support](https://docs.snapcraft.io/core/install) on your system.

File: user/deployment/snaps.md [hunk 4]
```diff
@@ -1,38 +1,69 @@
...
+
+```bash
+sudo snap install snapcraft --classic
+```
+
+Now export a login covering the ability to push this specific snap, as well as release to the channel you selected (in this example, `edge`):
```
Comment (evandandrea) [inline user/deployment/snaps.md]: Login tokens can attenuate how, when, and where they can be used, thus minimising damage from compromise. For Travis, export a token that can only upload this snap to the channel you specified above (in this example, `edge`):

File: user/deployment/snaps.md [hunk 5]
```diff
@@ -1,38 +1,69 @@
...
 
-    ```yaml
-    deploy:
-      skip_cleanup: true
-      provider: script
-      script: docker run -v $(pwd):$(pwd) -w $(pwd) snapcore/snapcraft sh -c "apt update && snapcraft && snapcraft push *.snap --release edge"
-      on:
-        branch: master
-    ```
+(note the final `-`, which requests the login be exported to stdout instead of a file.) Copy the token, and put it into the `$SNAP_TOKEN` environment variable in Travis CI:
```
Comment (evandandrea) [inline user/deployment/snaps.md]: _Note: The final `-` requests the login be exported to stdout instead of a file. It is required._

The token will be printed out. Copy and put it into the Travis CI environment variable `$SNAP_TOKEN`:

File: user/deployment/snaps.md [hunk 6]
```diff
@@ -1,38 +1,69 @@
...
 
-The `snapcraft enable-ci travis` command will assist you getting the credential, encrypting it and adding the right script to the deploy section of your `.travis.yml` file. Here is a [tutorial that will guide you setting up the continuous delivery from Travis CI](https://tutorials.ubuntu.com/tutorial/continuous-snap-delivery-from-travis-ci#0) on your project.
+```bash
+travis env set SNAP_TOKEN "<token>"
+```
+
+If you use a different environment variable, you'll need to specify the `token` option for the provider:
```
Comment (evandandrea) [inline user/deployment/snaps.md]: I think we should drop this. Let's keep them on rails.

File: user/deployment/snaps.md [hunk 7]
```diff
@@ -1,38 +1,69 @@
...
+deploy:
+  provider: snap
+  snap: my.snap
+  channel: edge
+  token: $MY_SNAP_TOKEN
+```
+{: data-file=".travis.yml"}
 
-Note that the `edge` channel is intended for crowdtesting with your community of early adopters. With this deployment script in Travis, every time a pull request lands into the master branch, a new snap will be published to edge. Your testers can install it in any of the supported Linux distros with:
+This example is almost certainly not ideal, as you probably want to upload a snap built during CI. Set `skip_cleanup` to `true` to prevent Travis CI from deleting it before the `deploy` step.
```
Comment (evandandrea) [inline user/deployment/snaps.md]: I don't understand why we're skipping cleanup here.

We should find a better way of guiding them to how they \[build snaps](https://docs.snapcraft.io/build-snaps/languages) than saying "this is not ideal."

File: user/deployment/snaps.md [hunk 8]
```diff
@@ -1,38 +1,69 @@
...
+```yaml
+deploy:
+  provider: snap
+  snap: my.snap
+  channel: edge
+  skip_cleanup: true
+```
+{: data-file=".travis.yml"}
+
+Note that the `edge` channel is intended for crowdtesting with your community of early adopters. With this deployment setup in Travis, every time a pull request lands, a new snap will be published to edge. Your testers can install it in any of the supported Linux distros with:
```
Comment (evandandrea) [inline user/deployment/snaps.md]: Note: The `edge` channel is intended for the bleeding edge: your every commit to master will be built and uploaded. This is great for crowd-testing with your community of early adopters, who can install your app in any of the \[supported Linux distributions](https://docs.snapcraft.io/core/install) with:
```bash
sudo snap install my-snap-name --edge
```

Each upload gets a monotonically increasing integer. When you're ready, you can release one of these built commits to the stable channel for the public to discover in the \[Snap storefront](https://snapcraft.io/store). For example, you could promote the very first upload to stable:
```bash
snapcraft release my-snap-name 1 edge
```

File: user/deployment/snaps.md [hunk 9]
```diff
@@ -1,38 +1,69 @@
...
 
 ```bash
 sudo snap install my-snap-name --edge
 ```
 
 After that, they will be always ready to provide early feedback and help making a more stable release.
```
Comment (evandandrea) [inline user/deployment/snaps.md]: Drop

File: user/deployment/snaps.md [hunk 10]
```diff
@@ -1,38 +1,69 @@
...
 
-You can adjust the Travis script and the `snapcraft` calls to fit your development process. For example, you could install the snap and run user acceptance tests before releasing to edge, to make sure that there are no regressions in master. Also, there are three other channels in the store: `beta`, `candidate` and `stable`. You can run different tests suites with Travis for each channel on your way to stable.
-
 There are plenty of [tutorials](https://tutorials.ubuntu.com/) and [videos](https://www.youtube.com/snapcraftio) that will help you packaging your project as a snap.
```
Comment (evandandrea) [inline user/deployment/snaps.md]: Drop.

File: user/deployment/snaps.md [hunk 11]
```diff
@@ -1,38 +1,69 @@
 ---
-title: Ubuntu Snap Store
+title: Snap Store
 layout: en
 permalink: /user/deployment/snaps/
 ---
 
-Travis CI can automate the continuous delivery to the [Snaps](https://snapcraft.io) Store.
+Travis CI can automatically upload and release your built snap to the [snap store](https://snapcraft.io) after a successful build.
```
[the remark this answers is not in the document]
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md (hunk 1 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md (hunk 2 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md (hunk 3 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md (hunk 4 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md (hunk 5 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md (hunk 6 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Sure, although I'm not sure how they'll learn it's an option this way. Done.

File: user/deployment/snaps.md (hunk 7 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md (hunk 8 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md (hunk 9 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md (hunk 10 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Done.

File: user/deployment/snaps.md [hunk 12]
```diff
@@ -1,38 +1,52 @@
 ---
-title: Ubuntu Snap Store
+title: Snap Store
 layout: en
 permalink: /user/deployment/snaps/
 ---
 
-Travis CI can automate the continuous delivery to the [Snaps](https://snapcraft.io) Store.
+Travis CI can automatically upload and release your snap to the [snap store](https://snapcraft.io) after a successful build.
```
Comment (evandandrea) [inline user/deployment/snaps.md]: Snap Store is capitalised. I'd say release your project or release your app.

File: user/deployment/snaps.md [hunk 13]
```diff
@@ -1,38 +1,52 @@
...
+Note: The `edge` channel is intended for the bleeding edge: your every commit to master will be built and uploaded. This is great for crowd-testing with your community of early adopters, who can install your app in any of the [supported Linux distributions](https://docs.snapcraft.io/core/install) with:
 
-There are plenty of [tutorials](https://tutorials.ubuntu.com/) and [videos](https://www.youtube.com/snapcraftio) that will help you packaging your project as a snap.
+```bash
+sudo snap install my-snap-name --edge
+```
+
+Each upload gets a monotonically increasing integer. When you're ready, you can release one of these built commits to the stable channel for the public to discover in the [Snap storefront](https://snapcraft.io/store). For example, you could promote the very first upload to stable:
```
Comment (evandandrea) [inline user/deployment/snaps.md]: Sorry, I know I said "the public to discover", but what do you think of "for public discovery in the"?

File: user/deployment/snaps.md [hunk 14]
```diff
@@ -1,38 +1,52 @@
...
+_Note: The final `-` requests the login be exported to stdout instead of a file. It is required._
+
+The token will be printed out. Copy and put it into the Travis CI environment variable `$SNAP_TOKEN`:
+
+```bash
+travis env set SNAP_TOKEN "<token>"
+```
 
-You can adjust the Travis script and the `snapcraft` calls to fit your development process. For example, you could install the snap and run user acceptance tests before releasing to edge, to make sure that there are no regressions in master. Also, there are three other channels in the store: `beta`, `candidate` and `stable`. You can run different tests suites with Travis for each channel on your way to stable.
+Note: The `edge` channel is intended for the bleeding edge: your every commit to master will be built and uploaded. This is great for crowd-testing with your community of early adopters, who can install your app in any of the [supported Linux distributions](https://docs.snapcraft.io/core/install) with:
```
Comment (evandandrea) [inline user/deployment/snaps.md]: _Note: The `edge` channel is intended for the bleeding edge: your every commit to master will be built and uploaded._

Your community of early-adopters and testers can install your app in any of the [supported Linux distributions](https://docs.snapcraft.io/core/install) with:

File: user/deployment/snaps.md (hunk 12 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: I'm a tad worried that changing "snap" to "app" will make folks assume the deployer will build their snap for them, but done.

File: user/deployment/snaps.md (hunk 13 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Sure, done.

File: user/deployment/snaps.md (hunk 14 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Very good, done.

File: user/deployment/snaps.md [hunk 15]
```diff
@@ -1,38 +1,54 @@
...
 
-    ```yaml
-    deploy:
-      skip_cleanup: true
-      provider: script
-      script: docker run -v $(pwd):$(pwd) -w $(pwd) snapcore/snapcraft sh -c "apt update && snapcraft && snapcraft push *.snap --release edge"
-      on:
-        branch: master
-    ```
+In order for Travis CI to have permission to upload the snap on your behalf, you need to export a Snap Store login token. You can do this with the snapraft command-line tool, once you have [enabled snap support](https://docs.snapcraft.io/core/install) on your system.
```
Comment (plaindocs) [inline user/deployment/snaps.md]: s/snapraft/snapcraft/

Comment (plaindocs) [reply user/deployment/snaps.md]: Maybe reword the first part to

> To uploads snaps from Travis CI, export a ...

File: user/deployment/snaps.md [hunk 16]
```diff
@@ -1,38 +1,54 @@
...
 
-The `snapcraft enable-ci travis` command will assist you getting the credential, encrypting it and adding the right script to the deploy section of your `.travis.yml` file. Here is a [tutorial that will guide you setting up the continuous delivery from Travis CI](https://tutorials.ubuntu.com/tutorial/continuous-snap-delivery-from-travis-ci#0) on your project.
+```bash
+sudo snap install snapcraft --classic
+```
 
-Note that the `edge` channel is intended for crowdtesting with your community of early adopters. With this deployment script in Travis, every time a pull request lands into the master branch, a new snap will be published to edge. Your testers can install it in any of the supported Linux distros with:
+Login tokens can attenuate how, when, and where they can be used, thus minimising damage from compromise. For Travis CI, export a token that can only upload this snap to the channel you specified above (in this example, `edge`):
```
Comment (plaindocs) [inline user/deployment/snaps.md]: Might be worth using something other than attenuate, for clarity, what do you think?

File: user/deployment/snaps.md (hunk 15 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Good catch, thank you. Fixed both.

File: user/deployment/snaps.md (hunk 16 above)
Comment (kyrofa) [reply user/deployment/snaps.md]: Agreed, the word "specify" probably makes more sense, here. Fixed.

**Comment by plaindocs** [comment]
Thanks @kyrofa for persevering with all of the reviews.
